File: code_reality/tour_manifest.py
"""corpus provenance manifest——.tours/manifest.toml 讀寫。

derived/curated 二分的機械載體：source×generator×anchored_commit。
curated＝generator "manual"；重產 diff 非空的 derived 由 audit 建議升 manual（不覆蓋）。
"""

import argparse
import json
import math
import re
import subprocess
from pathlib import Path

try:
    import tomllib
except ModuleNotFoundError:  # pragma: no cover
    tomllib = None  # type: ignore[assignment]


def git_head(repo: Path) -> str:
    out = subprocess.run(
        ["git", "-C", str(repo), "rev-parse", "HEAD"],
        capture_output=True,
        text=True,
        check=False,
    )
    if out.returncode != 0:
        print(
            f"[WARN] git HEAD 取不到（{repo} 非 git repo？）——anchored_commit 記 unknown"
        )
        return "unknown"
    return out.stdout.strip()


def load(path: Path) -> dict:
    if not path.exists():
        return {}
    if tomllib is None:  # pragma: no cover
        raise RuntimeError("需要 Python 3.11+（tomllib）")
    return tomllib.loads(path.read_text(encoding="utf-8"))


def upsert(
    data: dict,
    rel: str,
    *,
    generator: str,
    sources: list[str],
    commit: str,
) -> dict:
    rows = data.setdefault("tour", {})
    rows[rel] = {
        "generator": generator,
        "sources": sources,
        "anchored_commit": commit,
    }
    return data


def _kv(key: str, val: str) -> str:
    return f'{key} = "{val}"'


def _toml_key(key: str) -> str:
    """bare key（[A-Za-z0-9_-]+）直出；其餘 quote——鍵名裸輸出非 bare 形態會寫出不可解析 TOML。"""
    if re.fullmatch(r"[A-Za-z0-9_-]+", key):
        return key
    return json.dumps(key, ensure_ascii=False)


def _toml_value(v: object) -> str:
    """頂層未知鍵的 TOML 序列化（scalar／scalar list）；非支援型別 loud——silent 掉資料更糟。"""
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, float) and not math.isfinite(v):
        raise ValueError(
            "manifest 頂層鍵含非有限 float（inf/nan）——TOML 無此字面，寫出即非法"
        )
    if isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, str):
        # json 跳脫是 TOML basic string 子集；U+007F 兩家語義不同——補跳脫
        # （否則寫出非法 TOML、延遲到下游 load 才炸）
        return json.dumps(v, ensure_ascii=False).replace("\x7f", "\\u007f")
    if isinstance(v, list) and all(isinstance(x, (bool, int, float, str)) for x in v):
        return "[" + ", ".join(_toml_value(x) for x in v) + "]"
    raise ValueError(
        f"manifest 頂層鍵型別不支援保存（{type(v).__name__}）——只支援 scalar／scalar list"
    )


def dump(path: Path, data: dict) -> None:
    lines = [f"version = {_toml_value(data.get('version', 1))}"]
    # 未知頂層鍵 roundtrip 保存——dump 只重建已知欄位會 silent 刪除人工鍵
    # （F7：NT 的 audience = "newcomer" 兩次被 upsert 刪掉）
    for key in sorted(k for k in data if k not in ("version", "tour")):
        lines.append(f"{_toml_key(key)} = {_toml_value(data[key])}")
    for rel in sorted(data.get("tour", {})):
        row = data["tour"][rel]
        lines.append(f'\n[tour."{rel}"]')
        lines.append(_kv("generator", row["generator"]))
        srcs = ", ".join(f'"{s}"' for s in row.get("sources", []))
        lines.append(f"sources = [{srcs}]")
        lines.append(_kv("anchored_commit", row["anchored_commit"]))
        # row 未知鍵同原則保存（J1）——upsert 全列替換＝重產列歸工具權威；
        # 未動列 roundtrip 不得掉資料
        for key in sorted(
            k for k in row if k not in ("generator", "sources", "anchored_commit")
        ):
            lines.append(f"{_toml_key(key)} = {_toml_value(row[key])}")
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")


def tours_root_of(out_dir: Path) -> Path:
    """out_dir（如 .tours/arch/<stem>）往上找名為 .tours 的根；找不到一路上至 filesystem root——呼叫端以 ``name != ".tours"`` 判定非 corpus 樹。"""
    p = out_dir.resolve()
    while p.name and p.name != ".tours" and p.parent != p:
        p = p.parent
    return p


def init_scan(
    repo: Path,
    tours_dir: Path,
    *,
    generator_rule: str = "chain",
) -> dict:
    """掃 corpus 補 manifest——只補缺行（既有行不覆蓋：generator 原生寫入的 sources 保留）；generator 以檔名慣例猜（`chain-*` 或純序號 `NN.tour`→chain_tour、其餘 manual）、sources 留空。"""
    path = repo / tours_dir / "manifest.toml"
    data = load(path) if path.exists() else {}
    data.setdefault("version", 1)
    data.setdefault("tour", {})
    commit = git_head(repo)
    for f in sorted((repo / tours_dir).rglob("*.tour")):
        rel_path = f.relative_to(repo / tours_dir)
        if any(d in rel_path.parts[:-1] for d in ("delta", "dev-fixture")):
            continue  # 時間層可再生／開發假資料——非 manifest 範圍
        rel = rel_path.as_posix()
        if rel in data["tour"]:
            continue
        gen = (
            "chain_tour"
            if generator_rule == "chain"
            and (f.name.startswith("chain-") or (f.stem.isascii() and f.stem.isdigit()))
            else "manual"
        )
        upsert(data, rel, generator=gen, sources=[], commit=commit)
    return data


def main() -> None:
    parser = argparse.ArgumentParser(description="manifest 讀寫／--init-scan 骨架生成")
    parser.add_argument("--repo", type=Path, default=Path.cwd())
    parser.add_argument("--tours-dir", type=Path, default=Path(".tours"))
    parser.add_argument(
        "--init-scan", action="store_true", help="掃 corpus 生成 manifest 骨架"
    )
    args = parser.parse_args()
    path = args.repo / args.tours_dir / "manifest.toml"
    if not args.init_scan:
        print(f"[OK] manifest path: {path}（exists={path.exists()}）")
        return
    data = init_scan(args.repo, args.tours_dir)
    dump(path, data)
    print(f"[OK] manifest init: {len(data['tour'])} rows -> {path}")


if __name__ == "__main__":
    main()

File: crates/code-reality/src/tour_manifest.rs
//! `tour_manifest` — the frozen `code_reality/tour_manifest.py` contract:
//! `.tours/manifest.toml` read/write (the derived/curated split as
//! source×generator×anchored_commit). dump roundtrips UNKNOWN keys —
//! rebuilding only the known fields would silently delete hand-written
//! entries (the NT `audience` incident, twice).

use crate::argparse::{parse, FlagSpec, Kind, Outcome, ToolSpec};
use crate::ToolOutput;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

const SPEC: ToolSpec = ToolSpec {
    flags: &[
        FlagSpec {
            long: "--repo",
            short: None,
            kind: Kind::Value { metavar: "REPO" },
        },
        FlagSpec {
            long: "--tours-dir",
            short: None,
            kind: Kind::Value {
                metavar: "TOURS_DIR",
            },
        },
        FlagSpec {
            long: "--init-scan",
            short: None,
            kind: Kind::StoreTrue,
        },
    ],
    positionals: &[],
};

const HELP: &str = concat!(
    "usage: tour_manifest [-h] [--repo REPO] [--tours-dir TOURS_DIR]\n",
    "                     [--init-scan]\n",
    "\n",
    "manifest 讀寫／--init-scan 骨架生成\n",
    "\n",
    "options:\n",
    "  -h, --help            show this help message and exit\n",
    "  --repo REPO\n",
    "  --tours-dir TOURS_DIR\n",
    "  --init-scan           掃 corpus 生成 manifest 骨架\n",
);

/// git HEAD with the WARN fallback (`tour_manifest.py:20-32`): non-git
/// repo → "[WARN] ... anchored_commit 記 unknown" on stdout, "unknown".
pub fn git_head(repo: &Path) -> (String, String) {
    let out = std::process::Command::new("git")
        .arg("-C")
        .arg(repo)
        .arg("rev-parse")
        .arg("HEAD")
        .output();
    match out {
        Ok(o) if o.status.success() => (
            String::from_utf8_lossy(&o.stdout).trim().to_string(),
            String::new(),
        ),
        _ => (
            "unknown".to_string(),
            format!(
                "[WARN] git HEAD 取不到（{} 非 git repo？）——anchored_commit 記 unknown\n",
                repo.display()
            ),
        ),
    }
}

/// Manifest data model: `version`, `tour` (rel → row), plus unknown
/// top-level keys preserved for roundtripping.
#[derive(Debug, Clone, Default)]
pub struct Manifest {
    pub version: Option<toml::Value>,
    pub tour: BTreeMap<String, toml::Table>,
    pub extra: Vec<(String, toml::Value)>, // insertion-irrelevant: dump sorts
}

pub fn load(path: &Path) -> Result<Manifest, String> {
    if !path.exists() {
        return Ok(Manifest::default());
    }
    let text =
        std::fs::read_to_string(path).map_err(|e| format!("{} 讀取失敗：{}", path.display(), e))?;
    let data: toml::Value =
        toml::from_str(&text).map_err(|e| format!("{} TOML 解析失敗：{}", path.display(), e))?;
    let table = data.as_table().cloned().unwrap_or_default();
    let mut m = Manifest::default();
    for (k, v) in table {
        match k.as_str() {
            "version" => m.version = Some(v),
            "tour" => {
                if let Some(t) = v.as_table() {
                    for (rel, row) in t {
                        if let Some(row_t) = row.as_table() {
                            m.tour.insert(rel.clone(), row_t.clone());
                        }
                    }
                }
            }
            other => m.extra.push((other.to_string(), v)),
        }
    }
    Ok(m)
}

pub fn upsert(m: &mut Manifest, rel: &str, generator: &str, sources: &[String], commit: &str) {
    let mut row = toml::Table::new();
    row.insert(
        "generator".into(),
        toml::Value::String(generator.to_string()),
    );
    row.insert(
        "sources".into(),
        toml::Value::Array(
            sources
                .iter()
                .map(|s| toml::Value::String(s.clone()))
                .collect(),
        ),
    );
    row.insert(
        "anchored_commit".into(),
        toml::Value::String(commit.to_string()),
    );
    m.tour.insert(rel.to_string(), row);
}

fn toml_key(key: &str) -> String {
    // bare key ([A-Za-z0-9_-]+) verbatim; otherwise JSON-quoted
    let bare = !key.is_empty()
        && key
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-');
    if bare {
        key.to_string()
    } else {
        serde_json::to_string(&serde_json::Value::String(key.to_string())).unwrap()
    }
}

/// Scalar / scalar-list TOML serialization (`tour_manifest.py:71-89`);
/// unsupported types are loud — silently dropping data is worse.
fn toml_value(v: &toml::Value) -> Result<String, String> {
    match v {
        toml::Value::Boolean(b) => Ok(b.to_string()),
        toml::Value::Integer(i) => Ok(i.to_string()),
        toml::Value::Float(f) => {
            if !f.is_finite() {
                return Err(
                    "manifest 頂層鍵含非有限 float（inf/nan）——TOML 無此字面，寫出即非法"
                        .to_string(),
                );
            }
            Ok(format!("{}", f))
        }
        toml::Value::String(s) => {
            // json escaping is a TOML basic-string subset; U+007F differs
            // between the two — escape it or the file is unparseable
            Ok(serde_json::to_string(&serde_json::Value::String(s.clone()))
                .unwrap()
                .replace('\u{7f}', "\\u007f"))
        }
        toml::Value::Array(items) => {
            let mut parts = Vec::new();
            for x in items {
                match x {
                    toml::Value::Boolean(_)
                    | toml::Value::Integer(_)
                    | toml::Value::Float(_)
                    | toml::Value::String(_) => parts.push(toml_value(x)?),
                    _ => {
                        return Err(
                            "manifest 頂層鍵型別不支援保存（list 內非 scalar）——只支援 scalar／scalar list"
                                .to_string(),
                        );
                    }
                }
            }
            Ok(format!("[{}]", parts.join(", ")))
        }
        other => Err(format!(
            "manifest 頂層鍵型別不支援保存（{:?}）——只支援 scalar／scalar list",
            other.type_str()
        )),
    }
}

pub fn dump(path: &Path, m: &Manifest) -> Result<(), String> {
    let mut lines = Vec::new();
    let version = m.version.clone().unwrap_or(toml::Value::Integer(1));
    lines.push(format!("version = {}", toml_value(&version)?));
    // unknown top-level keys roundtrip (sorted)
    let mut extra = m.extra.clone();
    extra.sort_by(|a, b| a.0.cmp(&b.0));
    for (k, v) in &extra {
        lines.push(format!("{} = {}", toml_key(k), toml_value(v)?));
    }
    for (rel, row) in &m.tour {
        lines.push(format!("\n[tour.\"{rel}\"]"));
        lines.push(format!(
            "generator = \"{}\"",
            row.get("generator").and_then(|v| v.as_str()).unwrap_or("")
        ));
        let srcs: Vec<String> = row
            .get("sources")
            .and_then(|v| v.as_array())
            .map(|a| {
                a.iter()
                    .map(|s| format!("\"{}\"", s.as_str().unwrap_or("")))
                    .collect()
            })
            .unwrap_or_default();
        lines.push(format!("sources = [{}]", srcs.join(", ")));
        lines.push(format!(
            "anchored_commit = \"{}\"",
            row.get("anchored_commit")
                .and_then(|v| v.as_str())
                .unwrap_or("")
        ));
        // unknown row keys roundtrip (sorted) — untouched rows must not
        // lose data; upserted rows are tool-authoritative full replaces
        let mut unknown: Vec<(&String, &toml::Value)> = row
            .iter()
            .filter(|(k, _)| !matches!(k.as_str(), "generator" | "sources" | "anchored_commit"))
            .collect();
        unknown.sort_by(|a, b| a.0.cmp(b.0));
        for (k, v) in unknown {
            lines.push(format!("{} = {}", toml_key(k), toml_value(v)?));
        }
    }
    std::fs::write(path, format!("{}\n", lines.join("\n")))
        .map_err(|e| format!("{} 寫入失敗：{}", path.display(), e))
}

/// Walk up from out_dir to a `.tours` root (`tour_manifest.py:114-119`);
/// the caller judges "not a corpus tree" by `name != ".tours"`.
pub fn tours_root_of(out_dir: &Path) -> PathBuf {
    let mut p = crate::common::resolve(out_dir);
    while !p.file_name().unwrap_or_default().is_empty()
        && p.file_name().unwrap_or_default() != ".tours"
        && p.parent() != Some(&p)
    {
        p = p.parent().unwrap_or(&p).to_path_buf();
    }
    p
}

/// Scan the corpus filling ONLY missing rows (`tour_manifest.py:122-148`):
/// existing generator-written sources are preserved; generator guessed
/// from filename convention (`chain-*` or bare `NN.tour` → chain_tour).
pub fn init_scan(repo: &Path, tours_dir: &Path) -> Result<(Manifest, String), String> {
    let path = repo.join(tours_dir).join("manifest.toml");
    let mut data = load(&path)?;
    if data.version.is_none() {
        data.version = Some(toml::Value::Integer(1));
    }
    let (commit, warn) = git_head(repo);
    let root = repo.join(tours_dir);
    let mut files: Vec<PathBuf> = glob_tours(&root);
    files.sort();
    for f in files {
        let Ok(rel_path) = f.strip_prefix(&root) else {
            continue;
        };
        let parts: Vec<String> = rel_path
            .components()
            .map(|c| c.as_os_str().to_string_lossy().into_owned())
            .collect();
        let dirs = &parts[..parts.len().saturating_sub(1)];
        if dirs.iter().any(|d| d == "delta" || d == "dev-fixture") {
            continue; // regenerable time layer / dev fixtures — out of scope
        }
        let rel = rel_path.to_string_lossy().replace('\\', "/");
        if data.tour.contains_key(&rel) {
            continue;
        }
        let name = f
            .file_name()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_default();
        let stem = f
            .file_stem()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_default();
        let gen = if name.starts_with("chain-")
            || (stem.is_ascii() && stem.chars().all(|c| c.is_ascii_digit()))
        {
            "chain_tour"
        } else {
            "manual"
        };
        upsert(&mut data, &rel, gen, &[], &commit);
    }
    Ok((data, warn))
}

pub(crate) fn glob_tours(root: &Path) -> Vec<PathBuf> {
    let mut opts = glob::MatchOptions::new();
    opts.require_literal_leading_dot = false;
    let pattern = root.join("**/*.tour").to_string_lossy().into_owned();
    glob::glob_with(&pattern, opts)
        .into_iter()
        .flatten()
        .filter_map(|r| r.ok())
        .collect()
}

/// Route a `code-reality tour_manifest ...` invocation.
pub fn run(argv: &[&str]) -> ToolOutput {
    let Some((&_sub, toks)) = argv.split_first() else {
        return ToolOutput::fail("需提供子命令 tour_manifest");
    };
    let values = match parse(&SPEC, toks) {
        Outcome::Help => {
            return ToolOutput {
                stdout: HELP.to_string(),
                stderr: String::new(),
                exit_code: 0,
            };
        }
        Outcome::Err(msg) => return ToolOutput::fail(msg),
        Outcome::Ok { values, .. } => values,
    };
    let repo = values
        .get("--repo")
        .and_then(|v| v.clone())
        .map(PathBuf::from)
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")));
    let tours_dir = values
        .get("--tours-dir")
        .and_then(|v| v.clone())
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from(".tours"));
    let path = repo.join(&tours_dir).join("manifest.toml");
    if !values.contains_key("--init-scan") {
        return ToolOutput {
            stdout: format!(
                "[OK] manifest path: {}（exists={}）\n",
                path.display(),
                path.exists()
            ),
            stderr: String::new(),
            exit_code: 0,
        };
    }
    let (data, warn) = match init_scan(&repo, &tours_dir) {
        Ok(v) => v,
        Err(e) => return ToolOutput::crash(e),
    };
    if let Err(e) = dump(&path, &data) {
        return ToolOutput::crash(e);
    }
    ToolOutput {
        stdout: format!(
            "{}[OK] manifest init: {} rows -> {}\n",
            warn,
            data.tour.len(),
            path.display()
        ),
        stderr: String::new(),
        exit_code: 0,
    }
}
